vowpalwabbit/reductions/epsilon_decay.cc

```diff
@@ -0,0 +1,256 @@
+      break;
+    }
+  }
+
+  // check if any model counts are higher than the champion. If so, shift the model
+  // back to the beginning of the list and reset its counts
+  auto model_idx = model_count - 1;
+  for (auto candidate_iter = champion_iter + 1; candidate_iter != end_iter; ++candidate_iter, --model_idx)
+  {
+    if (candidate_iter->update_count > data._min_scope &&
+        candidate_iter->update_count >
+            (std::pow(champion_iter->update_count, static_cast<float>(model_idx) / model_count)))
+    {
+      auto n_iter = swap_models(candidate_iter + 1, candidate_iter, end_iter);
+      reset_models(n_iter, end_iter, data._weights, data._epsilon_decay_alpha, data._epsilon_decay_tau, model_count);
+      break;
+    }
+  }
+}
+
+void save_load_epsilon_decay(
+    VW::reductions::epsilon_decay::epsilon_decay_data& epsilon_decay, io_buf& io, bool read, bool text)
+{
+  if (io.num_files() == 0) { return; }
+  if (read) { VW::model_utils::read_model_field(io, epsilon_decay); }
+  else
+  {
+    VW::model_utils::write_model_field(io, epsilon_decay, "_epsilon_decay", text);
+  }
+}
+
+VW::LEARNER::base_learner* VW::reductions::epsilon_decay_setup(VW::setup_base_i& stack_builder)
+{
+  VW::config::options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+
+  std::string arg;
+  bool epsilon_decay_option;
+  uint64_t model_count;
+  uint64_t _min_scope;
+  float _epsilon_decay_alpha;
+  float _epsilon_decay_tau;
+
+  option_group_definition new_options("Epsilon-Decaying Exploration");
+  new_options
+      .add(make_option("epsilon_decay", epsilon_decay_option)
+               .necessary()
+               .keep()
+               .help("Use decay of exploration reduction"))
+      .add(make_option("model_count", model_count).keep().default_value(3).help("Set number of exploration models"))
+      .add(make_option("min_scope", _min_scope)
+               .keep()
+               .default_value(100)
+               .help("Minimum example count of model before removing"))
+      .add(make_option("epsilon_decay_alpha", _epsilon_decay_alpha)
+               .keep()
+               .default_value(DEFAULT_ALPHA)
+               .help("Set confidence interval for champion change"))
+      .add(make_option("epsilon_decay_tau", _epsilon_decay_tau)
+               .keep()
+               .default_value(DEFAULT_TAU)
+               .help("Time constant for count decay"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) { return nullptr; }
+
+  // Update model count to be 2^n
+  uint64_t params_per_weight = 1;
+  while (params_per_weight < model_count) { params_per_weight *= 2; }
+
+  // Scale confidence interval by number of examples
+  float scaled_alpha = _epsilon_decay_alpha / model_count;
+
+  auto data = VW::make_unique<VW::reductions::epsilon_decay::epsilon_decay_data>(
+      model_count, _min_scope, scaled_alpha, _epsilon_decay_tau, all.weights);
```

Comment:
Checks for sane model_count?  perhaps > 0?  @peter pointed out that iterators later in the code make this assumption.

Reply:
added check